src/shc.cpp

```diff
@@ -0,0 +1,292 @@
+// Doth Libray (C) by Marcos Oliveira
+
+#include <doht/shc/shc.h>
+
+#include <cstddef>
+#include <functional>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <optional>
+#include <queue>
+#include <stack>
+#include <string>
+#include <vector>
+
+namespace doht::shc {
+
+namespace {
+struct Node {
+  std::optional<char> symbol;
+  size_t frequency;
+};
+
+class CodingTree {
+ public:
+  void ClearTree() {
+    weights_.clear();
+    code_map_.clear();
+  }
+
+  void BuildTree(const std::map<char, int32_t>& frequencies) {
+    if (weights_.size() || code_map_.size()) {
+      ClearTree();
+    }
+
+    SetSize(frequencies.size());
+
+    struct HeapNode : Node {
+      HeapNode(const std::optional<char>& symbol, size_t freq) {
+        this->symbol = symbol;
+        this->frequency = freq;
+      }
+
+      std::shared_ptr<HeapNode> left{};
+      std::shared_ptr<HeapNode> right{};
+
+      bool operator>(const HeapNode& rhs) {
+        return this->frequency > rhs.frequency;
+      }
+    };
+
+    class SharedHeapNodeFunctor {
+     public:
+      bool operator()(const std::shared_ptr<HeapNode>& lhs,
+                      const std::shared_ptr<HeapNode>& rhs) {
+        if (!lhs.get()) {
+          return false;
+        }
+
+        if (!rhs.get()) {
+          return true;
+        }
+
+        return *lhs > *rhs;
+      }
+    };
+
+    auto frequency_to_node =
+        [](const std::pair<char, int32_t>& entry) -> std::shared_ptr<HeapNode> {
+      return std::make_shared<HeapNode>(std::make_optional(entry.first),
+                                        static_cast<size_t>(entry.second));
+    };
+
+    std::vector<std::shared_ptr<HeapNode>> leafs{};
+    leafs.reserve(frequencies.size());
+    for (auto& it : frequencies) {
+      leafs.emplace_back(frequency_to_node(it));
+    }
+
+    std::priority_queue<std::shared_ptr<HeapNode>,
+                        std::vector<std::shared_ptr<HeapNode>>,
+                        SharedHeapNodeFunctor>
```

Comment:
Since there is only one occurrence, perhaps you could use std::function and pass a lambda in the constructor of the priority queue